src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,910 @@
+    private static final int NTLM_AUTHENTICATE_PAYLOAD_OFFSET = 88;
+
+    // client challenge length
+    private static final int NTLM_CLIENT_NONCE_LENGTH = 8;
+
+    // server challenge length
+    private static final int NTLM_SERVER_CHALLENGE_LENGTH = 8;
+
+    // Windows Filetime timestamp length
+    private static final int NTLM_TIMESTAMP_LENGTH = 8;
+
+    // Windows epoch time difference from Unix epoch time in secs
+    private static final long WINDOWS_EPOCH_DIFF = 11644473600L;
+
+    /**
+     * NTLM Client Context
+     */
+    private class NTLMContext {
+        // domain name to authenticate to
+        private final String domainName;
+
+        // unicode bytes of domain name
+        private final byte[] domainUbytes;
+
+        // user credentials
+        private final String upperUserName;
+        private final byte[] userNameUbytes;
+        private final byte[] passwordHash;
+
+        // workstation
+        private String workstation;
+
+        // unicode bytes of server SPN
+        private final byte[] spnUbytes;
+
+        // message authentication code
+        private Mac mac = null;
+
+        // negotiate flags from Challenge msg
+        private long negotiateFlags = 0x00000000;
+
+        // session key calculated from password
+        private byte[] sessionBaseKey = null;
+
+        // Windows FileTime timestamp - number of 100 nanosecond ticks since Windows Epoch time
+        private byte[] timestamp = null;
+
+        // target info field from Challenge msg
+        private byte[] targetInfo = null;
+
+        // server challenge from Challenge msg
+        private byte[] serverChallenge = new byte[NTLM_SERVER_CHALLENGE_LENGTH];
+
+        /**
+         * Section 3.1.5.1.2 Client Receives a CHALLENGE_MESSAGE from the Server
+         * 
+         * Save messages for calculating MIC
+         */
+        private byte[] negotiateMsg = null;
+        private byte[] challengeMsg = null;
+
+        /**
+         * Creates an NTLM client context
+         *
+         * @param con
+         *        connection to SQL server
+         * @param domainName
+         *        domain name to authentication in using NTLM
+         * @param userName
+         *        user name
+         * @param password
+         *        password
+         * @param workstation
+         *        hostname of the workstation
+         * @throws SQLServerException
+         *         if error occurs
+         */
+        NTLMContext(final SQLServerConnection con, final String domainName, final String userName,
+                final String password, final String workstation) throws SQLServerException {
+
+            this.domainName = null != domainName ? domainName.toUpperCase()
+                                                 : SQLServerDriverStringProperty.DOMAIN.getDefaultValue();
+            this.domainUbytes = unicode(this.domainName);
+
+            this.userNameUbytes = null != userName ? unicode(userName) : null;

```

Comment:
The `null` case is not properly handled here, if `userName` is `null`, the next `this.upperUserName = userName.toUpperCase();` line will result in `NullPointerException`.